src/api/push.rs

```diff
@@ -0,0 +1,297 @@
+            debug!("Device doesn't exist");
+            return;
+        }
+    };
+
+    let data = json!({
+        "userId": user_uuid,
+        "organizationId": (),
+        "deviceId": device.push_uuid,
+        "identifier": acting_device_uuid,
+        "type": ut as i32,
+        "payload": {
+            "Id": cipher.uuid,
+            "UserId": cipher.user_uuid,
+            "OrganizationId": (),
+            "RevisionDate": cipher.updated_at
+        }
+    });
+
+    send_to_push_relay(data).await;
+}
+
+pub async fn push_logout(user: &User, acting_device_uuid: Option<String>, conn: &mut crate::db::DbConn) {
+    let data: JsonValue;
+    if let Some(d) = acting_device_uuid {
+        if let Some(device) = Device::find_by_uuid_and_user(d.as_str(), &user.uuid, conn).await {
+            data = json!({
+                "userId": user.uuid,
+                "organizationId": (),
+                "deviceId": device.push_uuid,
+                "identifier": d,
+                "type": UpdateType::LogOut as i32,
+                "payload": {
+                    "UserId": user.uuid,
+                    "Date": user.updated_at
+                }
+            });
+            send_to_push_relay(data).await;
+        } else {
+            debug!("Device doesn't exist");
+        }
+    } else {
+        data = json!({
+            "userId": user.uuid,
+            "organizationId": (),
+            "deviceId": (),
+            "identifier": (),
+            "type": UpdateType::LogOut as i32,
+            "payload": {
+                "UserId": user.uuid,
+                "Date": user.updated_at
+            }
+        });
+        send_to_push_relay(data).await;
+    }
+}
+
+pub async fn push_user_update(ut: UpdateType, user: &User) {
+    let data = json!({
+        "userId": user.uuid,
+        "organizationId": (),
+        "deviceId": (),
+        "identifier": (),
+        "type": ut as i32,
+        "payload": {
+            "UserId": user.uuid,
+            "Date": user.updated_at
+        }
+    });
+
+    send_to_push_relay(data).await;
+}
+
+pub async fn push_folder_update(
+    ut: UpdateType,
+    folder: &Folder,
+    acting_device_uuid: &String,
+    conn: &mut crate::db::DbConn,
+) {
+    let device = match Device::find_by_uuid_and_user(acting_device_uuid, &folder.user_uuid, conn).await {
+        Some(d) => d,
+        None => {
+            debug!("Device doesn't exist");
+            return;
+        }
+    };
+
+    let data = json!({
+        "userId": folder.user_uuid,
+        "organizationId": (),
+        "deviceId": device.push_uuid,
+        "identifier": acting_device_uuid,
+        "type": ut as i32,
+        "payload": {
+            "Id": folder.uuid,
+            "UserId": folder.user_uuid,
+            "RevisionDate": folder.updated_at
+        }
+    });
+
+    send_to_push_relay(data).await;
+}
+
+pub async fn push_send_update(ut: UpdateType, send: &Send) {
+    if send.user_uuid.is_none() {
```

Comment:
Maybe a good idea too also add the check if push is enabled or not?

```suggestion
    if !CONFIG.push_enabled() || send.user_uuid.is_none() {
```
That prevents the code from continuing and in the end still being aborted.

Reply:
Already done [here](https://github.com/dani-garcia/vaultwarden/blob/4ec5334a77d471dd8f6082f482b80b785e4f65e0/src/api/push.rs#L271)